libqpdf/QPDFCrypto_mbedtls.cc

```diff
@@ -0,0 +1,187 @@
+#include <qpdf/QPDFCrypto_mbedtls.hh>
+
+#ifdef QPDF_MBEDTLS_2
+# include <mbedtls/entropy_poll.h>
+#else
+# include <mbedtls/entropy.h>
+#endif
+
+#include <mbedtls/error.h>
+#include <stdexcept>
+
+static void
+check_mbedtls(int status)
+{
+    if (status < 0) {
+        char err[128];
+        mbedtls_strerror(status, err, sizeof(err));
+        throw std::runtime_error(std::string("MbedTLS error: ") + err);
+    };
+}
+
+static void
+bad_bits(int bits)
+{
+    throw std::logic_error(
+        std::string("MbedTLS error: unsupported key length ") + std::to_string(bits));
+}
+
+QPDFCrypto_mbedtls::QPDFCrypto_mbedtls()
+{
+    memset(md_out, 0, sizeof(md_out));
+    mbedtls_md_init(&md_ctx);
+    mbedtls_cipher_init(&cipher_ctx);
+}
+
+QPDFCrypto_mbedtls::~QPDFCrypto_mbedtls()
+{
+    mbedtls_md_free(&md_ctx);
+    mbedtls_cipher_free(&cipher_ctx);
```

Comment:
Freeing these two contexts unconditionally is a safe operation, as far as I can tell.